### PR TITLE
fix detect end of literal

### LINKED ISSUE #1637: Compiler hangs on returning postfixed number literal from a function with mismatching types
[Reproduction link](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPWhhbGYKZXhwb3J0IGZ1bmN0aW9uIGJ1ZygpOiBpMzIgewogIGNvbnN0IHggPSAxdTg7CgogIHJldHVybiB4Owp9CiMhaHRtbAo8dGV4dGFyZWEgaWQ9Im91dHB1dCIgc3R5bGU9ImhlaWdodDogMTAwJTsgd2lkdGg6IDEwMCUiIHJlYWRvbmx5PjwvdGV4dGFyZWE+CjxzY3JpcHQ+CmxvYWRlci5pbnN0YW50aWF0ZShtb2R1bGVfd2FzbSwgeyAvKiBpbXBvcnRzICovIH0pCiAgLnRoZW4oKHsgZXhwb3J0cyB9KSA9PiB7CiAgICBjb25zdCBvdXRwdXQgPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgnb3V0cHV0JykKICAgIGZvciAobGV0IGkgPSAwOyBpIDw9IDEwOyArK2kpIHsKICAgICAgb3V0cHV0LnZhbHVlICs9IGBmaWIoJHtpfSkgPSAke2V4cG9ydHMuZmliKGkpfVxuYAogICAgfQogIH0pCjwvc2NyaXB0Pg==).
```ts
export function bug(): i32 {
  const x = 1u8;

  return x;
}
```
The compiler hangs when using a postfixed literal (does ASC even have these?), but only if it's in a position other than the return value.

I.e., this compiles fine:
```ts
export function probably_still_a_bug(): i32 {
  return 1u8;
}
```

### PR DESCRIPTION
**Pull request by MaxGraey**
Possibly fix #1637 and #1125

- [x] I've read the contributing guidelines

### CODE REVIEW

File: src/util/text.ts [hunk 1]
```diff
@@ -175,6 +175,35 @@ export function isWhiteSpace(c: i32): bool {
   }
 }
 
+export function isEndOfLiteral(c: i32): bool {
+  switch (c) {
+    case CharCode.COMMA:
```
Comment (dcodeIO) [inline src/util/text.ts:180]: Exclusion lists like these make me a little nervous about forgetting a case. Wondering if there's something more precise we could do?

File: src/tokenizer.ts [hunk 2]
```diff
@@ -1467,6 +1486,12 @@ export class Tokenizer extends DiagnosticEmitter {
         }
         sepEnd = pos + 1;
       } else {
+        if (!isEndOfLiteral(c)) {
```
Comment (dcodeIO) [inline src/tokenizer.ts:1489]: Perhaps the checks can be moved to a place with less duplication. Perhaps into `testInteger`? Might have to extend it with more cases, though, hmm.

Comment (MaxGraey) [reply src/tokenizer.ts:1489]: Unfortunately it required a lot of refactorings and much more complex logic for `readInteger`

File: src/util/text.ts (hunk 1 above)
Comment (MaxGraey) [reply src/util/text.ts:180]: Unfortunate not. It also didn't fix cases like `5=;`, `6,;`, `5 7;`. All this required lookaheads

**Comment by MaxGraey** [comment]
It definitely need lookahead for cases like `5 insanceof i32` or `4 typeof ...`

**Pull request closed without merging**